File: src/main/java/dev/kxrim/HtmlBuilder.java
package dev.kxrim;

import dev.kxrim.elements.*;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardCopyOption;
import java.util.stream.Stream;

public class HtmlBuilder {
    private final StringBuilder html;
    private final StringBuilder head;
    private final String title;
    private final String lang;
    private final String charset;
    private final String viewport;
    private Theme currentTheme;
    private static final Path OUTPUT_DIR = Paths.get("generated");
    private static final Path ASSETS_DIR = OUTPUT_DIR.resolve("assets");

    public HtmlBuilder() {
        this("en", "UTF-8", "width=device-width, initial-scale=1.0", "Document");
    }

    public HtmlBuilder(String title) {
        this("en", "UTF-8", "width=device-width, initial-scale=1.0", title);
    }

    public HtmlBuilder(String lang, String charset, String viewport, String title) {
        html = new StringBuilder();
        head = new StringBuilder();

        this.lang = lang;
        this.charset = charset;
        this.viewport = viewport;
        this.title = title;

        buildHead();
    }

    private void buildHead() {
        head.setLength(0);
        head.append("<!DOCTYPE html>\n");
        head.append("<html lang=\"").append(lang).append("\">\n");
        head.append("<head>\n");
        head.append("<meta charset=\"").append(charset).append("\">\n");
        head.append("<meta name=\"viewport\" content=\"").append(viewport).append("\">\n");
        head.append("<title>").append(title).append("</title>\n");

        if (currentTheme != null) {
            head.append("<link rel=\"stylesheet\" href=\"assets/style.css\">\n");
        }

        head.append("</head>\n");
        head.append("<body>\n");
    }

    public HtmlBuilder addElement(Element element) {
        html.append(element.toHtml());
        return this;
    }

    // Fluent API methods
    public HtmlBuilder heading(int level, String text) {
        return addElement(new Heading(level, text));
    }

    public HtmlBuilder paragraph(String content) {
        return addElement(new Paragraph(content));
    }

    public HtmlBuilder div(String content) {
        return addElement(new Div(content));
    }

    public HtmlBuilder div(String content, String className) {
        return addElement(new Div(content, className));
    }

    public HtmlBuilder div(Div div) {
        return addElement(div);
    }

    public Div createDiv() {
        return new Div();
    }

    public Div createDiv(String className) {
        return new Div(className, true);
    }

    public HtmlBuilder button(String text) {
        return addElement(new Button(text));
    }

    public HtmlBuilder button(String text, String onClick) {
        return addElement(new Button(text, onClick));
    }

    public HtmlBuilder link(String url, String text) {
        return addElement(new Link(url, text));
    }

    public HtmlBuilder image(String src, String alt) {
        return addElement(new Image(src, alt));
    }

    public HtmlBuilder input(String type, String name, String placeholder) {
        return addElement(new Input(type, name, placeholder));
    }

    public HtmlBuilder textarea(String name, String placeholder, int rows, int cols) {
        return addElement(new Textarea(name, placeholder, rows, cols));
    }

    public HtmlBuilder listUnordered(String... items) {
        return addElement(new ListElement(false, items));
    }

    public HtmlBuilder listOrdered(String... items) {
        return addElement(new ListElement(true, items));
    }

    public HtmlBuilder blockQuote(String content) {
        return addElement(new BlockQuote(content));
    }

    public HtmlBuilder code(String content) {
        return addElement(new Code(content));
    }

    public HtmlBuilder strong(String content) {
        return addElement(new Strong(content));
    }

    public HtmlBuilder emphasis(String content) {
        return addElement(new Emphasis(content));
    }

    public HtmlBuilder horizontalRule() {
        return addElement(new HorizontalRule());
    }

    public HtmlBuilder useTheme(Theme theme) {
        this.currentTheme = theme;
        System.out.println("Theme set to: " + theme.getThemeName());
        return this;
    }

    private void applyTheme() {
        if (currentTheme == null) {
            return;
        }

        try {
            Files.createDirectories(ASSETS_DIR);

            String css = fetchThemeCssFromJson(currentTheme);

            if (css != null) {
                Path cssDestination = ASSETS_DIR.resolve("style.css");
                Files.writeString(cssDestination, css);
                System.out.println("Applied theme from remote: " + currentTheme.getThemeName());
            } else {
                Path themeCssSource = Paths.get(currentTheme.getStylePath());
                if (Files.exists(themeCssSource)) {
                    Path cssDestination = ASSETS_DIR.resolve("style.css");
                    Files.copy(themeCssSource, cssDestination, StandardCopyOption.REPLACE_EXISTING);
                    System.out.println("Applied theme from local files: " + currentTheme.getThemeName());
                } else {
                    System.err.println("Theme CSS not found locally or remotely");
                }
            }
        } catch (IOException e) {
            System.err.println("Error applying theme: " + e.getMessage());
        }
    }

    private String fetchThemeCssFromJson(Theme theme) {
        try {
            HttpClient client = HttpClient.newHttpClient();
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(Theme.getThemeJsonUrl()))
                    .GET()
                    .build();

            HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());

            if (response.statusCode() == 200) {
                String jsonResponse = response.body();
                return extractCssFromJson(jsonResponse, theme.name());
            }
        } catch (Exception e) {
            System.out.println("Could not fetch theme from remote, trying local files...");
        }
        return null;
    }

    private String extractCssFromJson(String json, String themeName) {
        try {
            String themeKey = "\"" + themeName + "\"";
            int themeStart = json.indexOf(themeKey);
            if (themeStart == -1) return null;

            int cssStart = json.indexOf("\"css\":", themeStart);
            if (cssStart == -1) return null;

            int cssValueStart = json.indexOf("\"", cssStart + 6) + 1;

            int cssValueEnd = cssValueStart;
            while (cssValueEnd < json.length()) {
                if (json.charAt(cssValueEnd) == '"' && json.charAt(cssValueEnd - 1) != '\\') {
                    break;
                }
                cssValueEnd++;
            }

            String cssContent = json.substring(cssValueStart, cssValueEnd);
            return cssContent.replace("\\n", "\n")
                           .replace("\\\"", "\"")
                           .replace("\\\\", "\\");
        } catch (Exception e) {
            System.err.println("Error parsing JSON: " + e.getMessage());
            return null;
        }
    }


    public HtmlBuilder copyAssets(String sourceDir) {
        try {
            Path sourcePath = Paths.get(sourceDir);
            if (!Files.exists(sourcePath) || !Files.isDirectory(sourcePath)) {
                System.out.println("Source directory does not exist: " + sourceDir);
                return this;
            }

            Files.createDirectories(ASSETS_DIR);

            try (Stream<Path> files = Files.walk(sourcePath)) {
                files.filter(Files::isRegularFile)
                     .filter(p -> {
                         String name = p.getFileName().toString().toLowerCase();
                         return name.endsWith(".png") || name.endsWith(".jpg") ||
                                name.endsWith(".jpeg") || name.endsWith(".gif") ||
                                name.endsWith(".svg") || name.endsWith(".webp");
                     })
                     .forEach(source -> {
                         try {
                             Path destination = ASSETS_DIR.resolve(source.getFileName());
                             Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING);
                             System.out.println("Copied asset: " + source.getFileName() + " -> assets/");
                         } catch (IOException e) {
                             System.err.println("Failed to copy: " + source);
                         }
                     });
            }
        } catch (IOException e) {
            System.err.println("Error copying assets: " + e.getMessage());
        }
        return this;
    }

    public HtmlBuilder addLocalImage(String localPath, String alt) {
        try {
            Path source = Paths.get(localPath);
            if (!Files.exists(source)) {
                System.err.println("Image file not found: " + localPath);
                addElement(new Image(localPath, alt)); // Add anyway with original path
                return this;
            }

            Files.createDirectories(ASSETS_DIR);
            String fileName = source.getFileName().toString();
            Path destination = ASSETS_DIR.resolve(fileName);
            Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING);

            addElement(new Image("assets/" + fileName, alt));
            System.out.println("Added local image: " + fileName);
        } catch (IOException e) {
            System.err.println("Error adding local image: " + e.getMessage());
            addElement(new Image(localPath, alt)); // Fallback to original path
        }
        return this;
    }

    public void build() {
        applyTheme();

        buildHead();

        head.append(html);
        head.append("</body>\n");
        head.append("</html>");

        try {
            Files.createDirectories(Paths.get("generated"));
            var filePath = Paths.get("generated/index.html");
            Files.writeString(filePath, head.toString());
            System.out.println("HTML file created at: " + filePath.toAbsolutePath());
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public String getLang() {
        return lang;
    }

    public String getCharset() {
        return charset;
    }

    public String getViewport() {
        return viewport;
    }

    public String getTitle() {
        return title;
    }
}

File: src/main/kotlin/team/htmlforge/HtmlBuilder.kt
package team.htmlforge

import team.htmlforge.elements.*
import java.io.IOException
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardCopyOption
import java.util.stream.Stream

class HtmlBuilder(
    private val lang: String = "en",
    private val charset: String = "UTF-8",
    private val viewport: String = "width=device-width, initial-scale=1.0",
    private val title: String = "Document"
) {
    private val html: StringBuilder = StringBuilder()
    private val head: StringBuilder = StringBuilder()
    private var currentTheme: Theme? = null

    var outputName: String = "index.html"

    companion object {
        val OUTPUT_DIR: Path = Paths.get("generated")
        val ASSETS_DIR: Path = OUTPUT_DIR.resolve("assets")
    }

    init {
        buildHead()
    }

    private fun buildHead() {
        head.setLength(0)
        head.append("<!DOCTYPE html>\n")
        head.append("<html lang=\"").append(lang).append("\">\n")
        head.append("<head>\n")
        head.append("<meta charset=\"").append(charset).append("\">\n")
        head.append("<meta name=\"viewport\" content=\"").append(viewport).append("\">\n")
        head.append("<title>").append(title).append("</title>\n")

        if (currentTheme != null) {
            head.append("<link rel=\"stylesheet\" href=\"assets/style.css\">\n")
        }

        head.append("</head>\n")
        head.append("<body>\n")
    }

    fun addElement(element: Element): HtmlBuilder {
        html.append(element.toHtml())
        return this
    }

    fun addText(text: String): HtmlBuilder {
        html.append(text)
        return this
    }

    // Fluent API methods
    fun heading(level: Int, text: String): HtmlBuilder = addElement(Heading(level, text))

    fun paragraph(content: String): HtmlBuilder = addElement(Paragraph(content))

    fun div(content: String): HtmlBuilder = addElement(Div(content))

    fun div(content: String, className: String): HtmlBuilder = addElement(Div(content, className))

    fun div(div: Div): HtmlBuilder = addElement(div)

    fun createDiv(): Div = Div()

    fun createDiv(className: String): Div = Div(className, true)

    fun button(text: String): HtmlBuilder = addElement(Button(text))

    fun button(text: String, onClick: String): HtmlBuilder = addElement(Button(text, onClick))

    fun link(url: String, text: String): HtmlBuilder = addElement(Link(url, text))

    fun image(src: String, alt: String): HtmlBuilder = addElement(Image(src, alt))

    fun input(type: String, name: String, placeholder: String): HtmlBuilder = addElement(Input(type, name, placeholder))

    fun textarea(name: String, placeholder: String, rows: Int, cols: Int): HtmlBuilder = addElement(Textarea(name, placeholder, rows, cols))

    fun listUnordered(vararg items: String): HtmlBuilder = addElement(ListElement(false, *items))

    fun listOrdered(vararg items: String): HtmlBuilder = addElement(ListElement(true, *items))

    fun blockQuote(content: String): HtmlBuilder = addElement(BlockQuote(content))

    fun code(content: String): HtmlBuilder = addElement(Code(content))

    fun strong(content: String): HtmlBuilder = addElement(Strong(content))

    fun emphasis(content: String): HtmlBuilder = addElement(Emphasis(content))

    fun horizontalRule(): HtmlBuilder = addElement(HorizontalRule())

    fun useTheme(theme: Theme): HtmlBuilder {
        this.currentTheme = theme
        println("Theme set to: ${theme.getThemeName()}")
        return this
    }

    private fun applyTheme() {
        val theme = currentTheme ?: return

        try {
            Files.createDirectories(ASSETS_DIR)

            val css = fetchThemeCssFromJson(theme)

            if (css != null) {
                val cssDestination = ASSETS_DIR.resolve("style.css")
                Files.writeString(cssDestination, css)
                println("Applied theme from remote: ${theme.getThemeName()}")
            } else {
                val themeCssSource = Paths.get(theme.getStylePath())
                if (Files.exists(themeCssSource)) {
                    val cssDestination = ASSETS_DIR.resolve("style.css")
                    Files.copy(themeCssSource, cssDestination, StandardCopyOption.REPLACE_EXISTING)
                    println("Applied theme from local files: ${theme.getThemeName()}")
                } else {
                    System.err.println("Theme CSS not found locally or remotely")
                }
            }
        } catch (e: IOException) {
            System.err.println("Error applying theme: ${e.message}")
        }
    }

    private fun fetchThemeCssFromJson(theme: Theme): String? {
        return try {
            val client = HttpClient.newHttpClient()
            val request = HttpRequest.newBuilder()
                .uri(URI.create(Theme.getThemeJsonUrl()))
                .GET()
                .build()

            val response = client.send(request, HttpResponse.BodyHandlers.ofString())

            if (response.statusCode() == 200) {
                val jsonResponse = response.body()
                extractCssFromJson(jsonResponse, theme.name)
            } else null
        } catch (e: Exception) {
            println("Could not fetch theme from remote, trying local files...")
            null
        }
    }

    private fun extractCssFromJson(json: String, themeName: String): String? {
        return try {
            val themeKey = "\"$themeName\""
            val themeStart = json.indexOf(themeKey)
            if (themeStart == -1) return null

            val cssStart = json.indexOf("\"css\":", themeStart)
            if (cssStart == -1) return null

            val cssValueStart = json.indexOf('"', cssStart + 6) + 1

            var cssValueEnd = cssValueStart
            while (cssValueEnd < json.length) {
                if (json[cssValueEnd] == '"' && json[cssValueEnd - 1] != '\\') {
                    break
                }
                cssValueEnd++
            }

            val cssContent = json.substring(cssValueStart, cssValueEnd)
            cssContent.replace("\\n", "\n")
                .replace("\\\"", "\"")
                .replace("\\\\", "\\")
        } catch (e: Exception) {
            System.err.println("Error parsing JSON: ${e.message}")
            null
        }
    }

    fun copyAssets(sourceDir: String): HtmlBuilder {
        try {
            val sourcePath = Paths.get(sourceDir)
            if (!Files.exists(sourcePath) || !Files.isDirectory(sourcePath)) {
                println("Source directory does not exist: $sourceDir")
                return this
            }

            Files.createDirectories(ASSETS_DIR)

            Files.walk(sourcePath).use { files ->
                files.filter { Files.isRegularFile(it) }
                    .filter { p ->
                        val name = p.fileName.toString().lowercase()
                        name.endsWith(".png") || name.endsWith(".jpg") ||
                                name.endsWith(".jpeg") || name.endsWith(".gif") ||
                                name.endsWith(".svg") || name.endsWith(".webp")
                    }
                    .forEach { source ->
                        try {
                            val destination = ASSETS_DIR.resolve(source.fileName)
                            Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING)
                            println("Copied asset: ${source.fileName} -> assets/")
                        } catch (e: IOException) {
                            System.err.println("Failed to copy: $source")
                        }
                    }
            }
        } catch (e: IOException) {
            System.err.println("Error copying assets: ${e.message}")
        }
        return this
    }

    fun addLocalImage(localPath: String, alt: String): HtmlBuilder {
        try {
            val source = Paths.get(localPath)
            if (!Files.exists(source)) {
                System.err.println("Image file not found: $localPath")
                addElement(Image(localPath, alt)) // Add anyway with original path
                return this
            }

            Files.createDirectories(ASSETS_DIR)
            val fileName = source.fileName.toString()
            val destination = ASSETS_DIR.resolve(fileName)
            Files.copy(source, destination, StandardCopyOption.REPLACE_EXISTING)

            addElement(Image("assets/$fileName", alt))
            println("Added local image: $fileName")
        } catch (e: IOException) {
            System.err.println("Error adding local image: ${e.message}")
            addElement(Image(localPath, alt)) // Fallback to original path
        }
        return this
    }

    fun build() {
        applyTheme()

        buildHead()

        head.append(html)
        head.append("</body>\n")
        head.append("</html>")

        try {
            Files.createDirectories(Paths.get("generated"))
            val filePath = Paths.get("generated/$outputName")
            Files.writeString(filePath, head.toString())
            println("HTML file created at: ${filePath.toAbsolutePath()}")
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    fun getLang() = lang
    fun getCharset() = charset
    fun getViewport() = viewport
    fun getTitle() = title
}
